man/POSIX/qstat.1
.TH "qstat" 1
.SH NAME
qstat \- query the queue manager for jobs
.SH SYNOPSIS
.B qstat
.I [--version] [-d] [-f] [-l] <jobid> <jobid>
.TP
.B qstat
.I [-d] -q <queue> <queue>
.SH DESCRIPTION
.PP
.B qstat
query the queue manager for all queued and active jobs. 
.SH OPTIONS
.TP
.B \-d \--debug
Turn on communication debugging
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-f \-\-full
Show more verbose information. 
.TP
.B \-\-header <header fields>
Specify custom header  with a colon-separated list for qstat display. (See qstat -lf for the possible fields.) The custom header may also be specified setting QSTAT_HEADER in the environment or a cqstat_header entry under [cqm] in /etc/cobalt.conf. The header is taken from /etc/cobalt.conf first, then $QSTAT_HEADER, and finally qstat --header.
.IP
.B Example:
.RS
.TP
qstat --header JobId:State:RunTime
.RE
.TP
.B \-l \-\-long
Show job info in vertical format. Use with '-f' to get all job info.
.TP
.B \-Q
Show queues and properties.
.IP
.B Properties:
.RS
.TP
.B MaxTime
The maximum walltime (in minutes) allowed for jobs submitted to the queue.
.TP
.B MinTime
The minimum walltime (in minutes) allowed for jobs submitted to the queue.
.TP
.B MaxQueued
The maximum number of jobs a user is allowed to have in the queue, whether running or queued.
.TP
.B Users
List of users allowed to submit to the queue. "None" means any user can submit to the queue.
.TP
.B State=running|stopped|draining|dead
Flag determining state of queue. If the queue is 'stopped', jobs can be submitted to the queue, but no jobs will be run by the scheduler. If the queue is 'draining', no new jobs will be allowed to be added to the queue, but jobs existing in the queue will be run. If the queue is 'dead', no new jobs will be added and no existing jobs will be run.
.TP
.B MaxRunning
The maximum number of jobs a user is allowed to have running in the queue.
.TP
.B MaxUsernodes
The maximum number of nodes a user is allowed to have allocated with running jobs in the queue.
.TP
.B TotalNodes
The maximum number of nodes any jobs in the queue are allowed to have allocated with running jobs.
.TP
.B Priority
The priority of the queue.  Higher numbers have higher priority.
.RE
.TP
.B \-\-reverse
Reverse ouput
.TP
.B \-\-sort <sort fields>
Specify sort fields  with a colon-separated list for qstat display.
.TP
.B \-u \-\-user <user name>
Specify user name.
.PP
.SH "ENIRONMENT VARIABLES"
The following environment variables affect qstat:
.TP
.B QSTAT_HEADER
Specifies custom header for job queries (See --header).
.SH "SEE ALSO"
.BR qdel (1),
.BR qsub (1),
.BR cqm (1)
.SH BUGS
None are currently known.


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 0.99.31 == 
 * Interactive mode jobs (qsub -I) are now supported on the BlueGene/Q.
   (Cobalt Ticket #107)
 * Timestamps are now included in cobaltlog lines (Cobalt Ticket #344 and #669)
 * Multiple mode arguments will result in a qsub rejection (Cobalt Ticket #664)
 * Fixed an issue that was causing COBALT_STARTTIME and COBALT_ENDTIME to not be
    set for compute node jobs in cX jobs. (Cobalt Ticket #673)

== Changes to 0.99.30 ==
 * This is a fix release that corrects a bug that can be encountered when upgrading
   to 0.99.27 from prior versions where ion_kerneloptions in cqm is improperly set.
   (Cobalt Ticket #665)

== Changes to 0.99.29 ==
 * Interactive jobs are experimentally available for cluster systems.

== Changes to 0.99.28 ==
 * Cobalt now supports qsub options in script mode jobs similarly to other
   job scheduler scripts.  The form is #Cobalt <qsub option> <option argument>.
   Most qsub options are supported, except for mode.  Jobs using this feature must
   be script mode jobs.  This is similar in style to PBS-directives. (Cobalt Ticket 354)
 * Fixed a bug in cluster systems where a failure duing job startup could result
   in all nodes for a job being stuck in the allocated state (Cobalt Ticket 662)
 * $COBALT_STARTTIME and $COBALT_ENDTIME added to job's environments, these are
   the Unix timestamp for when the job started, and when Cobalt will terminate
   the job due to exceeding walltime. (Cobat Ticket 593)
 * A requested walltime of 0 will set the job to run for the time remaining in
   a reservation. (Cobalt Ticket 474)
 * The --env flag to qsub now aggregates multiple instances of the flag rather
   than overriding the prior instance (Cobalt Ticket 657)

== Changes to 0.99.27 ==
 * Alternate kernel support for BG/Q compute nodes and IO nodes is available.
  Numerous configuration settings have been added to support this.
  (Cobalt ticket 621)
 * If $jobid is included in a job name specification or an environment variable,
  at submission time, Cobalt will substitute the generated jobid for that job.
  See man qsub for more details. (Cobalt ticket 617)
 * The Cobalt jobid is now explicitly added in the cobaltlog file. (Cobalt ticket 390)
 * The name of a job can now be specified seperately from naming the output
  files. (Cobalt ticket 625)
 * Multiple locations can be passed to reservations through the -p option as a
  ':'-delimited list. (Cobalt ticket 405)
 * cqadm can now be used to release user hold, as well as put jobs into a user hold
  as well as an admin hold (Cobalt ticket 382)
 * The users field can be removed from a reservation using setres -m -u '*'. (Cobalt
   ticket 381)
 * Releaseres should now report the correct partiton count on reservation release
  (Cobalt ticket 470)
 * Reservations now show time remaining in showres once the reservation is active.
  (Cobalt ticket 585)
 * Schedctl should handle flipping score and jobid arguments more gracefully and 
  no longer silently discards invalid options. (Cobalt ticket 502)
 * ':' and '=' may now be escaped as '\:' and '\=' in the --envs argument (Cobalt
  ticket 589)
 * Reservations may now take 'now' for the start time of a reservation (Cobalt
  ticket 546)
 * "GEOMETRY", "ION_KERNEL" and "ION_KERNELOPTIONS have been added to the JOB_DATA
  table of the Cobalt database 


== Changes to 0.99.26 ==
 * Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

== Changes to 0.99.25 ==
 * Fixes for clients that were causing erronious default values for proccount
  have been fixed.
 * The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
 * A bug that could cause TimeRemaining to show negative values has been fixed
 * Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


== Changes to 0.99.24 ==
 * Reverting client changes.  Found some issues when deploying to test systems
 * --defer flag has been reverted with these as a casualty.

== Changes to 0.99.23 ==
 * Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
 * IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
 * This release requires pybgsched-0.1.3
 * A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

== Changes to 0.99.22 ==
 * Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
 * Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
 * python 2.7 and it's change to the XMLRPC libraries are now supported.

== Changes to 0.99.21 ==
 * Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

== Changes to 0.99.20 ==
 * Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19
 * Better error logging for situations where the boots and the process group
  information haven't lined up yet.  This usually happens during high-load and
  can result in block boot completion never being detected.

== Changes to 0.99.19 ==
 * Manual boot control has been added
 * major refactor of compute block rebooting

== Changes to 0.99.18 ==
 * Fix for other systems that were seeing erronious geometry data
 * SoftwareFailure now accurately reported

== Changes to 0.99.17 ==

 * Support for --geometry has been added.  This will constrain a job to run 
blocks having a specific geometry in nodes, partadm, partlist and qstat have
been updated to show this.
 * Support for passthrough blocking reservations has been added.

src/clients/POSIX/qstat.py
#!/usr/bin/env python
"""
Cobalt Queue Status

Usage: %prog [options] <jobids1> ... <jobidsN>
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

Option with no values:

'-d','--debug',dest='debug',help='turn non communication debugging',callback=cb_debug
'-f','--full',dest='full',help='show more verbose information',action='store_true'
'-l','--long',dest='long',help='show job info in vertical format',action='store_true'
'-Q',dest='Q',help='show queues and properties',action='store_true'
'--reverse',dest='reverse',help='show output in reverse',action='store_true'

Option with values:

'--header',dest='header',type='string',help='specify custom header',callback=cb_split
'--sort',dest='sort',type='string',help='sort output by specified attribute',callback=cb_split
'-u','--user',dest='user',type='string',help='Specify username'

"""
import math
import os
import sys
import time
import types
import logging
from Cobalt import client_utils
from Cobalt.client_utils import \
    cb_debug, cb_split

from Cobalt.arg_parser import ArgParse

__revision__ = '$Revision: 406 $'
__version__ = '$Version$'

QUEMGR = client_utils.QUEMGR

def human_format(x):
    units = ['  ', ' K', ' M', ' G', ' T', ' P']
    dividend = 1000.0
    count = 0
    stuff = x

    while True:
        if stuff < dividend:
            return "%5.1f%s" % (max(stuff, 0.1), units[count])

        count += 1
        if count >= len(units):
            return "%5.1f%s" % (max(stuff, 0.1), units[-1])

        stuff = stuff / dividend

def get_output_for_queues(parser,hinfo):
    """
    get the queues info for the specified queues
    """

    names = parser.args if not parser.no_args() else ['*']

    query = [{'name':qname,'users':'*','mintime':'*','maxtime':'*','maxrunning':'*','maxqueued':'*','maxusernodes':'*',
              'maxnodehours':'*','totalnodes':'*','state':'*'} for qname in names]
    response = client_utils.component_call(QUEMGR, True, 'get_queues', (query,))

    if not parser.no_args() and not response:
        sys.exit(1)

    for q in response:
        if q['maxtime'] is not None:
            q['maxtime'] = "%02d:%02d:00" % (divmod(int(q['maxtime']), 60))
        if q['mintime'] is not None:
            q['mintime'] = "%02d:%02d:00" % (divmod(int(q['mintime']), 60))

    output = [[q[x] for x in [y.lower() for y in hinfo.header]] for q in response]

    return output

def get_output_for_jobs(parser,hinfo,queues):
    """
    get jobs from specified jobids
    """
    names              = parser.args if not parser.no_args() else ['*']
    user_name          = parser.options.user if parser.options.user != None else '*'
    query_dependencies = {'QueuedTime':
                              ['SubmitTime','StartTime'],'RunTime':['StartTime'],'TimeRemaining':['WallTime','StartTime']}

    try:
        query = []
        for n in names:
            if n=='*':
                query.append({'tag':'job', 'jobid':n, 'queue':'*'})
            elif [q['name'] for q in queues if q['name'] == n]:
                query.append({'tag':'job', 'queue':n, 'jobid':'*'})
            else:
                query.append({'tag':'job', 'jobid':int(n), 'queue':'*'})
    except ValueError:
        client_utils.logger.error("%s is not a valid jobid or queue name" % n)
        sys.exit(1)
    for q in query:
        for h in hinfo.long_header:
            if h == 'JobName':
                q.update({'jobname':'*'})
            elif h not in ['JobID', 'Queue']:
                q.update({h.lower():'*'})
            if h in query_dependencies:
                for x in query_dependencies[h]:
                    if x not in hinfo.header:
                        q.update({x.lower():'*'})
        q["user"] = user_name

    response = client_utils.component_call(QUEMGR, False, 'get_jobs', (query,))

    if not parser.no_args() and not response:
        sys.exit(1)

    if response:
        maxjoblen = max([len(str(item.get('jobid'))) for item in response])
        jobidfmt = "%%%ss" % maxjoblen
    # calculate derived values
    for j in response:
        # walltime
        walltime_secs = int(j['walltime']) * 60
        t = int(float(j['walltime']))
        h = int(math.floor(t/60))
        t -= (h * 60)
        j['walltime'] = "%02d:%02d:00" % (h, t)
        # jobid
        j['jobid'] = jobidfmt % j['jobid']
        # location
        if isinstance(j['location'], types.ListType) and len(j['location']) > 1:
            j['location'] = client_utils.merge_nodelist(j['location'])
        elif isinstance(j['location'], types.ListType) and len(j['location']) == 1:
            j['location'] = j['location'][0]
        # queuedtime
        if j.get('starttime') in ('-1', 'BUG', 'N/A', None):
            j['queuedtime'] = client_utils.get_elapsed_time(float(j.get('submittime')), time.time())
        else:
            j['queuedtime'] = client_utils.get_elapsed_time(float(j.get('submittime')), float(j['starttime']))
        # runtime
        if j.get('starttime') in ('-1', 'BUG', 'N/A', None):
            j['runtime'] = 'N/A'
        else:
            currtime = time.time()
            j['runtime'] = client_utils.get_elapsed_time( float(j['starttime']), time.time())
        # starttime
        if j.get('starttime') in ('-1', 'BUG', None):
            j['starttime'] = 'N/A'
        else:
            orig_starttime = float(j['starttime'])
            j['starttime'] = client_utils.sec_to_str(float(j['starttime']))
        # timeremaining
        if j.get('starttime') in ['-1', 'BUG', 'N/A' ,None]:
            j['timeremaining'] = 'N/A'
        else:
            time_remaining = walltime_secs - (currtime - orig_starttime)
            if time_remaining < 0:
                j['timeremaining'] = '00:00:00'
            else:
                s = int(time_remaining) % 60
                m = (int(time_remaining) % 3600) / 60
                h = int(time_remaining) / 3600
                j['timeremaining'] = "%02d:%02d:%02d" % (h, m, s)
        # jobname
        outputpath = j.get('outputpath')
        jobname    = j.get('jobname')
        # envs
        if j['envs'] is None:
            j.update({'envs':''})
        else:
            j['envs'] = ' '.join([str(x) + '=' + str(y) for x, y in j['envs'].iteritems()])
        # args
        j['args'] = ' '.join(j['args'])

        # make the SubmitTime readable by humans
        j['submittime'] = client_utils.sec_to_str(float(j['submittime']))

        j['outputpath'] = outputpath
        j['errorpath'] = j.get('errorpath')
        j['user_list'] = ':'.join(j['user_list'])

        if j['geometry'] != None:
            j['geometry'] = "x".join([str(i) for i in j['geometry']])
        else:
            j['geometry'] = 'Any'
    # any header that was not present in the query response has value set to '-'
    output = [[j.get(x, '-') for x in [y.lower() for y in hinfo.header]]
              for j in response]

    return output

def process_the_output(output,parser,hinfo):
    """
    process the qstat output
    """
    fields            = ['score'] if parser.options.sort == None else [f.lower() for f in parser.options.sort]
    lower_case_header = [str(h).lower() for h in hinfo.header]
    idxes             = []
    
    for f in fields:
        try:
            idx = lower_case_header.index(f)
            idxes.append(idx)
        except:
            pass
    if not idxes:
        idxes.append(0)

    def _my_cmp(left, right):
        for idx in idxes:
            try:
                val = cmp(float(left[idx]), float(right[idx]))
            except:
                val = cmp(left[idx], right[idx])
            if val == 0:
                continue
            else:
                return val

        return 0

    output.sort(_my_cmp)

    if parser.options.reverse != None:
        output.reverse()
    
    if "short_state" in lower_case_header:
        idx = lower_case_header.index("short_state")
        hinfo.header[idx] = "S"

    if "score" in lower_case_header:
        idx = lower_case_header.index("score")
        for line in output:
            line[idx] = human_format(float(line[idx]))

    if parser.options.long != None:
        client_utils.print_vertical([tuple(x) for x in [hinfo.header] + output])
    else:
        client_utils.print_tabular([tuple(x) for x in [hinfo.header] + output])


def main():
    """
    qstat main
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    delim = ':'

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args (tuple) >
        ( cb_debug        , () ),
        ( cb_split        , (delim,) ) ]

    # Get the version information
    opt_def =  __doc__.replace('__revision__',__revision__)
    opt_def =  opt_def.replace('__version__',__version__)

    parser = ArgParse(opt_def,callbacks)
    parser.parse_it() # parse the command line

    # Get the header instance 
    hinfo = client_utils.header_info(parser)

    # Get the queues for job ids
    queues = client_utils.component_call(QUEMGR, True, 'get_queues', ([{'name':'*','state':'*'}],))

    #  if Q option specified then get the info for the specified queues 
    if parser.options.Q != None:

        output = get_output_for_queues(parser,hinfo)

    else:

        # build query from long_header (all fields) and fetch response        
        output = get_output_for_jobs(parser,hinfo,queues)

    process_the_output(output,parser,hinfo)
            
if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
